zarr/util.py

```diff
@@ -10,13 +11,121 @@
 from asciitree import BoxStyle, LeftAligned
 from asciitree.traversal import Traversal
 from collections.abc import Iterable
-from numcodecs.compat import ensure_ndarray, ensure_text
+from numcodecs.compat import ensure_text
 from numcodecs.registry import codec_registry
 from numcodecs.blosc import cbuffer_sizes, cbuffer_metainfo
 
 from typing import Any, Callable, Dict, Optional, Tuple, Union
 
 
+# TODO: move the function to numcodecs
+def ensure_ndarray(buf):
+    """Convenience function to coerce `buf` to numpy array-like, if it is not
+    already a numpy array-like.
+
+    Parameters
+    ----------
+    buf : array-like or bytes-like
+        A numpy array or any object exporting a buffer interface.
+
+    Returns
+    -------
+    arr : array-like
+        A numpy array-like array that shares memory with `buf`.
+
+    Notes
+    -----
+    This function will not create a copy under any circumstances, it is guaranteed to
+    return a view on memory exported by `buf`.
+    """
+
+    if isinstance(buf, np.ndarray):
+        # already a numpy array
+        arr = buf
+
+    elif isinstance(buf, array.array) and buf.typecode in 'cu':
+        # Guard condition, do not support array.array with unicode type, this is
+        # problematic because numpy does not support it on all platforms. Also do not
+        # support char as it was removed in Python 3.
+        raise TypeError('array.array with char or unicode type is not supported')
+
+    else:
+
+        # TODO: make it possible to register array-like types lazily.
+        try:
+            import cupy
+            if isinstance(buf, cupy.ndarray):
+                return buf
+        except ImportError:
+            pass
```

Comment:
Wondering if we can use entrypoints as @martindurant is doing in PR ( https://github.com/zarr-developers/numcodecs/pull/300 )

This would be good to do in conjunction with hooking in other compressors